File: go/mechanisms/evm/batch-settlement/facilitator/voucher.go
package facilitator

import (
	"context"
	"fmt"
	"math/big"

	x402 "github.com/x402-foundation/x402/go"
	"github.com/x402-foundation/x402/go/mechanisms/evm"
	batchsettlement "github.com/x402-foundation/x402/go/mechanisms/evm/batch-settlement"
	"github.com/x402-foundation/x402/go/types"
)

// VerifyVoucher verifies a batched voucher-only payload.
// Checks voucher signature, reads onchain channel state, validates cumulative ceiling.
func VerifyVoucher(
	ctx context.Context,
	signer evm.FacilitatorEvmSigner,
	payload *batchsettlement.BatchSettlementVoucherPayload,
	requirements types.PaymentRequirements,
	channelConfig batchsettlement.ChannelConfig,
) (*x402.VerifyResponse, error) {
	return verifyVoucherFields(ctx, signer, &payload.Voucher, channelConfig, requirements, false)
}

// VerifyRefundVoucher verifies a cooperative-refund payload's voucher.
// The voucher is zero-charge: maxClaimableAmount == chargedCumulativeAmount,
// which on a fresh channel may equal totalClaimed exactly.
func VerifyRefundVoucher(
	ctx context.Context,
	signer evm.FacilitatorEvmSigner,
	payload *batchsettlement.BatchSettlementRefundPayload,
	requirements types.PaymentRequirements,
	channelConfig batchsettlement.ChannelConfig,
) (*x402.VerifyResponse, error) {
	return verifyVoucherFields(ctx, signer, &payload.Voucher, channelConfig, requirements, true)
}

func verifyVoucherFields(
	ctx context.Context,
	signer evm.FacilitatorEvmSigner,
	voucher *batchsettlement.BatchSettlementVoucherFields,
	channelConfig batchsettlement.ChannelConfig,
	requirements types.PaymentRequirements,
	isRefund bool,
) (*x402.VerifyResponse, error) {
	if err := ValidateChannelConfig(channelConfig, voucher.ChannelId, requirements); err != nil {
		return nil, err
	}

	chainId, err := signer.GetChainID(ctx)
	if err != nil {
		return nil, x402.NewVerifyError(ErrChannelStateReadFailed, "", fmt.Sprintf("failed to get chain ID: %s", err))
	}

	valid, err := VerifyBatchedVoucherTypedData(
		ctx, signer,
		voucher.ChannelId,
		voucher.MaxClaimableAmount,
		channelConfig.PayerAuthorizer,
		channelConfig.Payer,
		voucher.Signature,
		chainId,
	)
	if err != nil {
		return nil, x402.NewVerifyError(ErrVoucherSignatureInvalid, channelConfig.Payer,
			fmt.Sprintf("voucher signature verification failed: %s", err))
	}
	if !valid {
		return nil, x402.NewVerifyError(ErrVoucherSignatureInvalid, channelConfig.Payer,
			"voucher signature is invalid")
	}

	state, err := ReadChannelState(ctx, signer, voucher.ChannelId)
	if err != nil {
		return nil, x402.NewVerifyError(ErrChannelStateReadFailed, channelConfig.Payer,
			fmt.Sprintf("failed to read channel state: %s", err))
	}

	// A non-existent or fully-drained channel reports balance==0 onchain
	if state.Balance.Sign() == 0 {
		return nil, x402.NewVerifyError(ErrChannelNotFound, channelConfig.Payer,
			fmt.Sprintf("channel %s not found or fully drained (balance=0)", voucher.ChannelId))
	}

	maxClaimable, ok := new(big.Int).SetString(voucher.MaxClaimableAmount, 10)
	if !ok {
		return nil, x402.NewVerifyError(ErrInvalidVoucherPayload, channelConfig.Payer,
			"invalid maxClaimableAmount")
	}

	// Refund vouchers are zero-charge and may equal totalClaimed; non-refund
	// vouchers must strictly increase claimable above totalClaimed.
	belowClaimed := false
	if isRefund {
		belowClaimed = maxClaimable.Cmp(state.TotalClaimed) < 0
	} else {
		belowClaimed = maxClaimable.Cmp(state.TotalClaimed) <= 0
	}
	if belowClaimed {
		return nil, x402.NewVerifyError(ErrMaxClaimableTooLow, channelConfig.Payer,
			fmt.Sprintf("maxClaimableAmount %s is below totalClaimed %s", maxClaimable.String(), state.TotalClaimed.String()))
	}

	if maxClaimable.Cmp(state.Balance) > 0 {
		return nil, x402.NewVerifyError(ErrMaxClaimableExceedsBal, channelConfig.Payer,
			fmt.Sprintf("maxClaimableAmount %s exceeds balance %s", maxClaimable.String(), state.Balance.String()))
	}

	return &x402.VerifyResponse{
		IsValid: true,
		Payer:   channelConfig.Payer,
		Extra:   BuildVerifyExtra(voucher.ChannelId, state),
	}, nil
}

File: typescript/packages/mechanisms/hedera/src/batch-settlement/facilitator/voucher.ts
import type { PaymentRequirements, VerifyResponse } from "@x402/core/types";
import type { FacilitatorHederaBatchSigner } from "../signer";
import type {
  BatchSettlementRefundPayload,
  BatchSettlementVoucherPayload,
  ChannelConfig,
} from "../types";
import * as Errors from "../errors";
import {
  readChannelState,
  signatureCheckErrorCode,
  validateChannelConfig,
  verifyVoucherSignature,
} from "./utils";

/**
 * Verifies a cumulative voucher payload against onchain channel state.
 *
 * @param signer - Facilitator signer used for onchain reads and signature verification.
 * @param payload - Voucher or refund payload with signed voucher fields.
 * @param requirements - Server payment requirements (asset, network, amount).
 * @param channelConfig - Reconstructed channel configuration for the payer/receiver pair.
 * @returns A {@link VerifyResponse} indicating validity and returning channel state in `extra`.
 */
export async function verifyVoucher(
  signer: FacilitatorHederaBatchSigner,
  payload: BatchSettlementVoucherPayload | BatchSettlementRefundPayload,
  requirements: PaymentRequirements,
  channelConfig: ChannelConfig,
): Promise<VerifyResponse> {
  const { voucher } = payload;
  const channelId = voucher.channelId;
  const payer = channelConfig.payer;

  const configErr = validateChannelConfig(channelConfig, channelId, requirements);
  if (configErr) {
    return { isValid: false, invalidReason: configErr, payer };
  }

  const sigCheck = await verifyVoucherSignature(
    signer,
    {
      channelId,
      maxClaimableAmount: voucher.maxClaimableAmount,
      payerAuthorizer: channelConfig.payerAuthorizer,
      payer: channelConfig.payer,
      signature: voucher.signature,
    },
    requirements.network,
  );
  if (!sigCheck.ok) {
    return {
      isValid: false,
      invalidReason: signatureCheckErrorCode(sigCheck, Errors.ErrInvalidVoucherSignature),
      invalidMessage: sigCheck.message,
      payer,
    };
  }

  let state;
  try {
    state = await readChannelState(signer, channelId, requirements.network);
  } catch (error) {
    return {
      isValid: false,
      invalidReason: Errors.ErrRpcReadFailed,
      invalidMessage: error instanceof Error ? error.message : String(error),
      payer,
    };
  }

  if (state.balance === 0n) {
    return { isValid: false, invalidReason: Errors.ErrChannelNotFound, payer };
  }

  const maxClaimableAmount = BigInt(voucher.maxClaimableAmount);

  if (maxClaimableAmount > state.balance) {
    return { isValid: false, invalidReason: Errors.ErrCumulativeExceedsBalance, payer };
  }

  const belowClaimed =
    payload.type === "refund"
      ? maxClaimableAmount < state.totalClaimed
      : maxClaimableAmount <= state.totalClaimed;
  if (belowClaimed) {
    return { isValid: false, invalidReason: Errors.ErrCumulativeAmountBelowClaimed, payer };
  }

  return {
    isValid: true,
    payer,
    extra: {
      channelId,
      balance: state.balance.toString(),
      totalClaimed: state.totalClaimed.toString(),
      withdrawRequestedAt: state.withdrawRequestedAt,
      refundNonce: state.refundNonce.toString(),
    },
  };
}
